rest_server/IOMgrConnection.go

```diff
@@ -0,0 +1,432 @@
+		buff := make([]byte, IOMgrChunkSize)
+		_, err = io.ReadFull(IOMgrConn, buff)
+
+		counter++
+	}
+
+}
+
+func (IOMgrConn *IOMgrConnection) streamJSONPayload(c *gin.Context) (err error) {
+
+	// Read Payload up to max json payload size
+	body := make([]byte, 0)
+	buffer := make([]byte, 1)
+	var bytesRead uint64 = 0
+
+	siodbLoggerPool.Debug("IOMgrConn.pool.maxJsonPayloadSize: %v", IOMgrConn.pool.maxJsonPayloadSize)
+
+	for true {
+		if bytesRead > IOMgrConn.pool.maxJsonPayloadSize {
+			break
+		}
+		newBytesRead, err := c.Request.Body.Read(buffer)
+		if newBytesRead == 0 {
+			if io.EOF == err {
+				break
+			} else if err != nil {
+				return err
+			}
+		}
+		body = append(body, buffer[0])
+		bytesRead = bytesRead + uint64(newBytesRead)
+	}
+	siodbLoggerPool.Debug("Request.body: %s", body)
+	siodbLoggerPool.Debug("bytes read: %v", bytesRead)
+	siodbLoggerPool.Debug("len(Request.body): %v", len(body))
+
+	var buf [binary.MaxVarintLen32]byte
+	// Write Payload length
+	encodedLength := binary.PutUvarint(buf[:], uint64(len(body)))
+	_, err = IOMgrConn.Write(buf[:encodedLength])
+	if nil != err {
+		return err
+	}
+
+	// Write raw payload
+	if _, err = IOMgrConn.Write(body); err != nil {
+		siodbLoggerPool.Error("err: %v", err)
+		return fmt.Errorf("not able to write payload to IOMgr: %v", err)
+	}
+
+	// Write 0 to indicate EOF
+	encodedLength = binary.PutUvarint(buf[:], uint64(0))
+	_, err = IOMgrConn.Write(buf[:encodedLength])
+	if nil != err {
+		siodbLoggerPool.Error("err: %v", err)
+	}
+
+	// Get Returned message
+	if err = IOMgrConn.readIOMgrResponse(true); err != nil {
+		if bytesRead > IOMgrConn.pool.maxJsonPayloadSize {
+			return fmt.Errorf("JSON payload truncated because it has bigger size than authoried (%v). IOMgr: %v",
+				IOMgrConn.pool.maxJsonPayloadSize, err)
+		}
+		return err
+	}
+
+	return nil
+
+}
+
+func (IOMgrConn *IOMgrConnection) writeIOMgrRequest(
+	restVerb SiodbIomgrProtocol.RestVerb,
+	objectType SiodbIomgrProtocol.DatabaseObjectType,
+	userName string,
+	token string,
+	objectName string,
+	objectId uint64) (err error) {
+
+	var databaseEngineRestRequest SiodbIomgrProtocol.DatabaseEngineRestRequest
+	databaseEngineRestRequest.RequestId = IOMgrConn.RequestID
+	databaseEngineRestRequest.Verb = restVerb
+	databaseEngineRestRequest.ObjectType = objectType
+
+	siodbLoggerPool.Debug("user: %v", userName)
+	databaseEngineRestRequest.UserName = userName
+	databaseEngineRestRequest.Token = token
+
+	if len(objectName) > 0 {
+		databaseEngineRestRequest.ObjectName = objectName
+	}
+	if objectId > 0 {
+		databaseEngineRestRequest.ObjectId = objectId
+	}
+	siodbLoggerPool.Debug("databaseEngineRestRequest: %v", databaseEngineRestRequest)
+
+	if _, err := IOMgrConn.writeMessage(DATABASEENGINERESTREQUEST, &databaseEngineRestRequest); err != nil {
+		return fmt.Errorf("Unable to write message to IOMgr: %v", err)
+	}
+
+	return nil
+
+}
+
+func (IOMgrConn *IOMgrConnection) readIOMgrResponse(incrementRequestID bool) (err error) {
```

Comment:
I think increment of request ID should happen in the write procedure. please get rid of this extra parameter.
it is better to pass here something like `expectedRequestId`